### PR TITLE
Dockerfile rewrite based on Ruby image with performance optimizations and size reduction, dedicated Streaming image

### PR DESCRIPTION
**Pull request by vmstan**
## Background

The current Dockerfile is reliant on an upstream container image `ghcr.io/moritzheiber/ruby-jemalloc` which while based on the official Ruby image, it is unpredictably updated, not tagged with any specific Debian version, and not controlled by the project. 

This problem was highlighted shortly before the 4.2 release, when the @moritzheiber package was recently inadvertently updated from Debian 11 to Debian 12 by an automation, without warning, which then broke our dependencies and its relationship to the second part of the Dockerfile which is the official Node image.

## Major Changes

- This PR creates a new dedicated Dockerfile for the Streaming API, and removes Node.js from the resulting core Mastodon image. 
- The main Dockerfile uses one container image for the base, the official Ruby image.
- The main Dockerfile now only loads Node.js (via official Node image) into the build phase in order to complete the `yarn bundle` and `assets:precompile` steps, but the resulting image **does not have Node installed**. 
- This uses the Ruby image as its main `base` layer, then spins off multiple `build` layers that adds all of the build tools, then does the `bundle` and `yarn` each in respective layers. This speeds up the process by allowing more tasks to run in parallel.
- The streaming Dockerfile uses one container image for the base, the official Node image.
- The streaming Dockerfile uses the recently merged ability of Yarn 4 to package only the required Node components for the streaming server resulting in a dramatically smaller image. (250MB, based on on the official Node image)
- It set the resulting `/opt/mastodon` directory to read only by the `mastodon` user which should be more secure (with the exception of the `system` subdirectory so it can be mounted for local files.

### Performance Changes

- Uses `patchelf` to enable libjemalloc2 only for Ruby applications (as suggested by @renchap)
- Applies performance enhancements for libjemalloc2 5.x (as suggested by @renchap)
- Enables Ruby [YJIT](https://shopify.engineering/ruby-yjit-is-production-ready) by default.

```
$ kubectl exec mastodon-5c75fdb874-52mnj -- ruby -v
ruby 3.2.2 (2023-03-30 revision e51014f9c0) +YJIT [x86_64-linux]
```

Left side shows non-optimized libjemalloc2 5.x memory usage from two Sidekiq/Mastodon Kubernetes hosts. Right side is after introducing memory optimizations.

CleanShot 2023-09-08 at 11 54 11@2x
CleanShot 2023-09-08 at 11 54 56@2x

### Available Build Arguments

There are arguments available for many new settings, which can be passed to build tools like `--build-arg MASTODON_VERSION_METADATA="glitch"`

- `MASTODON_VERSION_PRERELEASE=""` for support of overriding `beta3` or `dev.0` set by version.rb file
- `MASTODON_VERSION_METADATA=""`  for support of adding metadata to version string
- `NODE_MAJOR_VERSION="20"` for changing between major versions of Node repos (defaults to 20)
- `RAILS_SERVE_STATIC_FILES="true"` defaults to set Rails serving static files
- `RUBY_YJIT_ENABLE="1"` for enabling Ruby YJIT (set to 0 to disable, defaults to 1)
- `TZ="Etc/UTC"` for setting container image timezone
- `RUBY_VERSION="3.2.2"` for setting Ruby version
- `DEBIAN_VERSION="bookworm"` for setting Debian version

## Compile Comparison

|         | Time   | Size   |
|---------|--------|--------|
| Current | 155.7s | 1.47GB |
| This PR | 157.6s | 971MB  |

_Intel only, 4 CPU with no cache, best of three runs._

## Prior Work

This PR is a combination of many changes and techniques introduced in other work
- #23248
- #26861
- #25808
- #21137
- #25671
- #24702
- #21031
- #23552
- #16244 

_This is very much a WIP and yet another entry into the "Dockerfile Thunderdome" but open to feedback and ways to improve size, speed and caching._ 

### Previous Testing

I have personally been running a modified version of the #23248 Dockerfile, which serves as the core of this version, on my instance for the last two months. I am running these current Dockerfile on vmst.io for regular builds.

### CODE REVIEW

**Comment by vmstan** [comment]
Current build errors from GitHub actions are from missing changes coming as part of #26830

**Comment by shleeable** [comment]
Maybe something a little cleaner?

```
echo "deb [signed-by=/usr/share/keyrings/nodesource.gpg] https://deb.nodesource.com/node_${NODE_MAJOR_VERSION}.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list
echo "deb [signed-by=/usr/share/keyrings/yarnkey.gpg] https://dl.yarnpkg.com/debian stable main" | tee /etc/apt/sources.list.d/yarn.list

wget -nv -O - https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor | tee /usr/share/keyrings/nodesource.gpg >/dev/null
wget -nv -O - https://dl.yarnpkg.com/debian/pubkey.gpg | gpg --dearmor | tee /usr/share/keyrings/yarnkey.gpg >/dev/null

apt-get update
apt-get install -y --no-install-recommends ffmpeg nodejs yarn

if [ "${DEBIAN_MM_REPO}" = "1" ]; then
    DEB_MM_REPO_URL="https://mirror.csclub.uwaterloo.ca/debian-multimedia/"
    echo "deb [signed-by=/usr/share/keyrings/deb-multimedia-keyring.gpg] ${DEB_MM_REPO_URL} bookworm main non-free" | tee /etc/apt/sources.list.d/deb-mm.list
    wget -nv -O - https://www.deb-multimedia.org/pool/main/d/deb-multimedia-keyring/deb-multimedia-keyring_2016.8.1_all.deb | dpkg-deb -x - ./
    gpg --dearmor ./etc/apt/trusted.gpg.d/deb-multimedia-keyring.gpg
    mv ./etc/apt/trusted.gpg.d/deb-multimedia-keyring.gpg /usr/share/keyrings/deb-multimedia-keyring.gpg >/dev/null
    apt-get install -y --no-install-recommends imagemagick-7
else
    apt-get install -y --no-install-recommends imagemagick
fi

```

**Comment by renchap** [comment]
Welcome to the Dockerfile Thunderdome :D

I indeed would like to rewrite the Dockerfile, I am not sure we want to risk doing it before 4.2 is released, but definitely for 4.3.

Some initial remarks:

> * `DEBIAN_MM_REPO="0"` for enabling the Debian Multimedia repo (adds ImageMagick 7 and ffmpeg 6)

I am not sure I like this, I would prefer to stick to official Debian packages rather than externally maintained ones without any security guarantees.

> * `BIND="0.0.0.0"` for setting bind address

Do you have an example of where this would be useful? As this is running in a container, you should listen to 0.0.0.0, and have your container engine expose the ports.

> * `RAILS_ENV="production"` for setting Rails environment
> * `NODE_ENV="production"` for setting Node environment

I do not think it would make any sense to have those with other values than `production`, and it will probably wont work with the container at all.

> * `RUBY_VERSION="3.2.2"` for setting Ruby version
> * `DEBIAN_VERSION="bookworm"` for setting Debian version

I am also not 100% convinced we should enable `jemalloc` using `LD_PRELOAD`, as this will also enable it for any other binary we launch, for example ImageMagick & ffmpeg, which might not be what we want.

An alternative would be to use `patchelf` ([idea taken from here](https://github.com/docker-library/ruby/issues/182#issuecomment-1368432448)), like this:

```
apt-get install -yq --no-install-recommends libjemalloc2 patchelf && patchelf --add-needed libjemalloc.so.2 /usr/local/bin/ruby 
```

Also note that your example is using `jemalloc` 5, which according to some articles does not perform as well as `jemalloc` 3  regarding memory usage. Configuring jemalloc should help in this regard, and on the Rails Performance Slack someone reported that this config makes version 5 use less memory that version 3, while being faster:

```
ENV MALLOC_CONF='abort_conf: true, narenas:2, background_thread: true, thp: never, dirty_decay_ms: 1000, muzzy_decay_ms: 0'
```

Ideally this should require some testing, comparing the performance (response time) and memory, between your new image and the existing one.

**Comment by vmstan** [comment]
> Welcome to the Dockerfile Thunderdome :D

⚡ 

> I am not sure I like this, I would prefer to stick to official Debian packages rather than externally maintained ones without any security guarantees.

> Do you have an example of where this would be useful? As this is running in a container, you should listen to 0.0.0.0, and have your container engine expose the ports.

I do not. That came out of #25808 and even in writing the intro I couldn't figure out why I would want change this.

> I do not think it would make any sense to have those with other values than `production`, and it will probably wont work with the container at all.

Same as above. I have removed the ARG settings for both.
 
> I am also not 100% convinced we should enable `jemalloc` using `LD_PRELOAD`, as this will also enable it for any other binary we launch, for example ImageMagick & ffmpeg, which might not be what we want.

This is a good point and not something I'd considered.
 
> An alternative would be to use `patchelf` ([idea taken from here](https://github.com/docker-library/ruby/issues/182#issuecomment-1368432448)), like this:
> 
> ```
> apt-get install -yq --no-install-recommends libjemalloc2 patchelf && patchelf --add-needed libjemalloc.so.2 /usr/local/bin/ruby 
> ```

Just to verify would patchelf need to be executed every time the container is launched or does this persist?

> Also note that your example is using `jemalloc` 5, which according to some articles does not perform as well as `jemalloc` 3 regarding memory usage. Configuring jemalloc should help in this regard, and on the Rails Performance Slack someone reported that this config makes version 5 use less memory that version 3, while being faster:
> 
> ```
> ENV MALLOC_CONF='abort_conf: true, narenas:2, background_thread: true, thp: never, dirty_decay_ms: 1000, muzzy_decay_ms: 0'
> ```
> 
> Ideally this should require some testing, comparing the performance (response time) and memory, between your new image and the existing one.

I will incorporate this change as well.

**Comment by vmstan** [comment]
> Maybe something a little cleaner? `untested`

Good point. I have them in as multiple IFs because I sort of added it at the last minute expecting some pushback on the idea generally, and wanted it optional. Putting it in a single block would make it easier to patch back in (even if just for myself) were the option completely removed.

**Comment by vmstan** [comment]
@renchap I get warnings from that MALLOC_CONF setting `<jemalloc>: Malformed conf string`

**Comment by vmstan** [comment]
### Build Stats
New
1.97GB x86_64 only
416sec x86_64 only

Current (Sept 8 in main)
1.51GB x86_64 only
396sec x86_64 only

As built with Docker Buildx on Intel i9 MacBook Pro
Colima with 8 vCPU and 8GB of RAM
`docker system purge --all` run prior to each build

Expelling unused components from the image is my next task.

**Comment by vmstan** [comment]
After created a dedicated build layer for the `bundle install` and `yarn install` components it's now:

1.56GB
389.2s

Just a hair larger than the existing Dockerfile.

**Comment by vmstan** [comment]
> I am not sure I like this, I would prefer to stick to official Debian packages rather than externally maintained ones without any security guarantees.

I have removed this completely.

**Comment by vmstan** [comment]
1.52GB
369.3s

**Comment by shleeable** [comment]
This is looking good as a stable Dockerfile. 

Debian running older releases is always an annoyance.. but it is what it is.

**Comment by vmstan** [comment]
I have just committed changes with the goal of having two dedicated Dockerfiles to break streaming out into it's own package.

The main Dockerfile now only loads Node.js into the build phase in order to complete the assets:precompile step, but the resulting image does not have Node installed. This reduces the size of the image to 1.3GB.

The streaming Dockerfile is currently based on Alpine Linux, and the official Node image. (This could be moved to Debian if desired.) 

Because of the way the Node components are bundled in Mastodon, Yarn has to install all of the packages to run both Streaming and the assets for the regular Mastodon image, but as these two components become seperated in the future this can be changed to further reduce the footprint of that image.

**Comment by vmstan** [comment]
Down to 915MB.

**Comment by vmstan** [comment]
Switched method of using Node in the build layer to match https://github.com/mastodon/mastodon/pull/26861

**Comment by vmstan** [comment]
Readied changes for #27073

**Comment by sinoru** [comment]
I think... It will be great importing commit history of called outed PRs, or squash and co-authored them.

**Comment by vmstan** [comment]
Yeah let's see what ends up happening, right now I'm working off feedback off dev team and experience deploying them. There are a ton of Dockerfile ideas out there. They all get squashed into a single commit from the PR. I grabbed the ImageMagick source compile out of your recent one to see what that would look like in production on vmst.io and its working, but I don't know what the core team wants to do in terms of managing that.

**Comment by vmstan** [comment]
Closes [#813 ](https://github.com/mastodon/mastodon/issues/813)

File: Dockerfile
Comment (renchap) [inline Dockerfile:216]: Same here

File: streaming/Dockerfile
Comment (renchap) [inline streaming/Dockerfile:40]: This should be able to re-use the same cache as above

File: streaming/Dockerfile
Comment (renchap) [inline streaming/Dockerfile:53]: And here as well

**Review by renchap — COMMENTED** [review]
Added some comments to add Docker cache volumes to speed up builds after the initial one.

File: Dockerfile
Comment (vmstan) [reply Dockerfile:101]: Yeah that was left over from trying to determine how to install an enhanced version. Consolidated now.

**Comment by vmstan** [comment]
Incorporated buildx caching

**Comment by vmstan** [comment]
Adds .dockerignore from #24702

**Comment by vmstan** [comment]
Removed all of the logic for detection of PNPM, Yarn Classic, etc.

File: Dockerfile
Comment (renchap) [inline Dockerfile:86]: By copying everything here, it means that if any file change, this layer will be invalidated and everything will be rebuilt.

Maybe we could avoid this, and copy:
- package.json + yarn.lock before running `yarn install`
- Gemfile* before running `bundle install`
- `.` before pre-compiling?

This would allow better cache re-use for builds where dependencies did not change.

File: Dockerfile
Comment (renchap) [inline Dockerfile:235]: Should we pre-compile bootsnap cache here as well, like in the Rails upstream Dockerfile?

File: streaming/Dockerfile
Comment (renchap) [inline streaming/Dockerfile:52]: I would prefer to not have wildcards here

Comment (vmstan) [reply streaming/Dockerfile:52]: Leftover from being able to test all the package managers. Will fix.

File: Dockerfile
Comment (vmstan) [reply Dockerfile:235]: Ah that must not have been committed while I was testing. I've had to replacement my dev box a couple times since we looked at that. Will fix.

**Comment by vmstan** [comment]
Updated sizing (as uploaded to Docker Hub)

**New Dockerfiles**
333.17MB Mastodon
194.54MB Streaming

**Old Dockerfile**
521.22MB All (Nightly)

**Review by renchap — COMMENTED** [review]
This looks good to me, but I would like to have another set of eyes on it.

Pinging @timetinytim, both for review and also because this switches to a separate container for the streaming server, and this will require changes for our deploys.

Which makes me think that we also need to adjust the Github CI workflows, with:
- building and pushing the two containers where we currently do it
- split the container image test workflows in two, to test both the app and the streaming build process (ideally triggering only the relevant one depending on the modified files)

**Comment by vmstan** [comment]
I had to add a second copy operation as the resulting image didn't actually have the source code the way that I had it yesterday 😳 Also tweaked layer names.

File: streaming/Dockerfile
Comment (ThisIsMissEm) [inline streaming/Dockerfile:25]: We should make sure that we explicitly set the PORT, rather than just relying on the default in the code of 4000.

**Review by ThisIsMissEm — CHANGES_REQUESTED** [review]
Just needs PORT to be explicitly set, otherwise if streaming changes it's port for whatever reason, the dockerfile won't break

File: streaming/Dockerfile
Comment (ThisIsMissEm) [reply streaming/Dockerfile:25]: @vmstan we also need `NODE_ENV=production` I think?

File: streaming/Dockerfile
Comment (ThisIsMissEm) [inline streaming/Dockerfile:65]: I think this could be done earlier maybe?

File: streaming/Dockerfile
Comment (ThisIsMissEm) [inline streaming/Dockerfile:73]: Is there a missing `CMD` here or something to actually start the streaming server?

Comment (vmstan) [reply streaming/Dockerfile:73]: We should totally have a default command 🤣 I set to manually in all my deployments but yeah it would be better for it to just be there.

File: streaming/Dockerfile
Comment (vmstan) [reply streaming/Dockerfile:65]: Earlier where?

File: streaming/Dockerfile
Comment (vmstan) [reply streaming/Dockerfile:25]: That would only be for building a new image with it set to something though right? When executed through Docker Compose or Kubernetes couldn't you set that as a runtime environment variable still?

Comment (vmstan) [reply streaming/Dockerfile]: It's set to expose port 4000 at the end, why would it also need to be here?

File: streaming/Dockerfile
Comment (ThisIsMissEm) [inline streaming/Dockerfile:96]: This would be better:

```suggestion
CMD [ node ./streaming/index.js ]
```

File: streaming/Dockerfile
Comment (ThisIsMissEm) [reply streaming/Dockerfile:81]: Actually, maybe the layer optimisation wouldn't make sense here, since line 69 copies the entire mastodon source, so you'd not actually be saving running this step if streaming only changed but corepack/yarn didn't

File: streaming/Dockerfile
Comment (ThisIsMissEm) [inline streaming/Dockerfile:34]:
```suggestion
  BIND="0.0.0.0" \
# Explicitly set PORT to match the exposed port
  PORT=4000 \
```

**Review by ThisIsMissEm — CHANGES_REQUESTED** [review]
To ensure that the port/expose is always in sync.

File: streaming/Dockerfile
Comment (ThisIsMissEm) [reply streaming/Dockerfile]: nah, it's basically that whilst streaming defaults to `4000` at the moment, if that ever changed, by being explicit with setting `PORT=4000` you'd always get the `EXPOSE` to match, otherwise you'd have say `EXPOSE 4000` but the server running on `:5000` and everything would break. (it is a bit of paranoia but setting PORT explicitly ensures it'll never break in that way

Comment (ThisIsMissEm) [reply streaming/Dockerfile]: Basically node.js doesn't know about `EXPOSE`, it only knows about the environment variables.

File: streaming/Dockerfile
Comment (vmstan) [reply streaming/Dockerfile:81]: It only copies the streaming specific packages. Not the entire source. It does that on the main Dockerfile for that build.

**Comment by vmstan** [comment]
Updated sizing, after incorporating #24702 as well

**New Dockerfiles**
1.01 GB Mastodon (394.57 MB compressed)
299 MB Streaming (91.16 MB compressed)

**Old Dockerfile**
1.47 GB All (Nightly) (521.22 MB compressed)

_Compressed metrics via Docker Hub_

**Comment by renchap** [comment]
The `package.json` split has been merged, can you update the PR with it?

**Comment by vmstan** [comment]
> The `package.json` split has been merged, can you update the PR with it?

Done

File: .github/workflows/build-nightly.yml
Comment (renchap) [inline .github/workflows/build-nightly.yml:55]: Should we still push it to Docker Hub? I remember the team mentioning that the `tootsuite` org on Docker Hub was legacy and deprecated, so we could only push to GHR

**Review by renchap — COMMENTED** [review]
Looks very good! Do you have the sizes with the latest changes?

Also I am wondering if we should have the `build-container-image` workflow build both images, so we do not need to update every action.

It could have a `push_to_images_streaming` attribute to set the names of the streaming image, and if present it builds it?

File: .github/workflows/build-nightly.yml
Comment (vmstan) [reply .github/workflows/build-nightly.yml:55]: Didn't they walk some of that back, or maybe it didn't apply to us like we thought?

**Comment by vmstan** [comment]
> Looks very good! Do you have the sizes with the latest changes?
> 
> Also I am wondering if we should have the `build-container-image` workflow build both images, so we do not need to update every action.
> 
> It could have a `push_to_images_streaming` attribute to set the names of the streaming image, and if present it builds it?



> Looks very good! Do you have the sizes with the latest changes?
> 
> Also I am wondering if we should have the `build-container-image` workflow build both images, so we do not need to update every action.
> 
> It could have a `push_to_images_streaming` attribute to set the names of the streaming image, and if present it builds it?

Streaming is down to 255MB
Main image is right at about 1GB.

**Comment by ClearlyClaire** [comment]
What's the upgrade path for admins already relying on docker?

Shouldn't `docker-compose.yml` be updated as well?

**Comment by vmstan** [comment]
> What's the upgrade path for admins already relying on docker?
> 
> Shouldn't `docker-compose.yml` be updated as well?

Yeah so Renchap and I had discussed this, should the file in `main` be changed to represent running the edge/nightly builds and then the release branches be updated to reflect the versioned tags? (Maybe with PR #27819) Right now `main` shows the 4.2.0 tag.

His concern is that folks would just go to the GitHub, grab the docker-compose.yml and go deploy the alpha version.

**Comment by timetinytim** [comment]
Finally gave this a look myself. I absolutely don't have the same knowledge and historical context as everyone else when it comes to changes like this, but from where I'm sitting, this looks really solid.

What I want to do is do a few quick tests in a staging environment with these Dockerfiles as a sanity check to make sure it holds up in practice. I will do that today and circle back in a little bit.

**Comment by vmstan** [comment]
@renchap what do you think about adding the build-image tag to run that workflow and generate usable images?

**Comment by renchap** [comment]
Not opposed to it, but it requires the branch to be on the main repo, so we will probably need to create a new PR to run them

**Review by timetinytim — APPROVED** [review]
I was able to confirm that the image works on staging, and was also able to test some of the git workflows locally on my machine, and everything looks OK as far as I can tell.

I say this is approved. Though I'll keep a lookout once it's merged to make sure all the workflows behave properly. :slightly_smiling_face:

**Comment by renchap** [comment]
Maybe let's wait next week for merging this?

**Review by renchap — APPROVED** [review]

File: Dockerfile
Comment (jippi) [inline Dockerfile:244]: @vmstan to use the apt cache mount, you need to `rm -f /etc/apt/apt.conf.d/docker-clean`, otherwise it will auto-clean after install, leaving you with an empty cache (as the first of the commands in this `RUN` group

Comment (vmstan) [reply Dockerfile:244]: That would explain a lot of the trouble I had with that when it was going through GitHub workflows.

It'll need to go into a new PR to fix.

File: Dockerfile
Comment (jippi) [inline Dockerfile:90]: I would recommend using `&&` over `;` in these multi-lines to ensure if one of them fail, the build fail with that single command, and not silently continues on with the new command (even if `-o pipefail` is being used) - will help avoid any future hurt if the `-o pipefail` is ever removed, intentionally or not.

File: Dockerfile
Comment (jippi) [inline Dockerfile:119]: @vmstan  micro-optimization:  move these up before the COPY of `package.json` and `.yarn` lines.

Will ensure that a change to any of those files won't invalidate the layer cache for these 2 copy statements. Which is nice for build machines, since it will use significantly less disk space over time due to reduced cache trashing

File: Dockerfile
Comment (jippi) [inline Dockerfile:228]: I wonder if it would be worth caching the temp files here ([blog post on the topic](https://www.bigbinary.com/blog/speeding-up-docker-image-build-process-of-a-rails-application#speeding-up-rake-assets-precompile-task-by-using-cache))

**Pull request merged**
